File: webserver/htdocs/api/dicom/posters.php
<?php
/**
 * Parse arbitrary multipart/form-data content
 * Note: null result or null values for headers or value means error
 * @return array|null [{"headers":array|null,"value":string|null}]
 * @param string|null $boundary
 * @param string|null $content
 */
function parse_multipart_content(?string $content, ?string $boundary): ?array {
  if(empty($content) || empty($boundary)) return null;
  $sections = array_map("trim", explode("--$boundary", $content));
  $parts = [];
  foreach($sections as $section) {
    if($section === "" || $section === "--") continue;
    $fields = explode("\r\n\r\n", $section);
    if(preg_match_all("/([a-z0-9-_]+)\s*:\s*([^\r\n]+)/iu", $fields[0] ?? "", $matches, PREG_SET_ORDER) === 2) {
      $headers = [];
      foreach($matches as $match) $headers[$match[1]] = $match[2];
    } else $headers = null;
    $parts[] = ["headers" => $headers, "value"   => $fields[1] ?? null];
  }
  return empty($parts) ? null : $parts;
}

// post one or more instances belonging to study; takes PatientID and PatientName from existing study
function poststow() {
  include 'config.php';
  $d=file_get_contents("php://input");
  $t = getallheaders();
  if (!array_key_exists("Content-Type", $t)) {
    echo 'Does not contain content type';
    return;
  }
  preg_match_all("/boundary=\"([^;= ]+)\"/", $t["Content-Type"], $r);
  $boundary= $r[1][0];

  $d = parse_multipart_content($d, $boundary);
  for ($i=0; $i<count($d); $i++) {
    $file = tmpfile();
    fwrite($file, $d[$i]["value"]);
    $path = stream_get_meta_data($file)['uri'];
    ob_start();
    passthru($exe . ' "--dolua:dofile([[posters.lua]]);poststow([['.$path.']])"');
    $var = $var . ob_get_contents();
    ob_end_clean();
    fclose($file);
  }

  header('Access-Control-Allow-Origin: *');
  header('Content-Type: application/json');
  echo $var;
}

// post arbitrary file to conquest attachfile: command
// script is in importConverter format; e.g. add lua: if needed
// scripts in the form-data take preference over query string
function attachfile($script) {
  include 'config.php';
  $file = null;
  $ext = '.dcm';
  
  if (array_key_exists('script', $_POST)) {
    $script = $_POST['script'];
  }

  if(count($_FILES) > 0){
    $first_file_index = $_FILES[array_keys($_FILES)[0]];
    $path = $first_file_index['tmp_name'];
    $ext = '.' . pathinfo($first_file_index['name'], PATHINFO_EXTENSION);
  }
  else {
    $d=file_get_contents("php://input");
    $t = getallheaders();
    if (!array_key_exists("Content-Type", $t)) {
      echo 'Does not contain content type';
      return;
    }
    $file = tmpfile();
    $typ=$t["Content-Type"];
    
    $ext = '.dcm';
    if (strpos($typ, 'application/zip')!==false) $ext='.zip';
    
    if (strpos($typ, 'multipart/related')!==false) {
      preg_match_all("/boundary=\"([^;= ]+)\"/", $typ, $r);
      $boundary= $r[1][0];
    
      $d = parse_multipart_content($d, $boundary);
      fwrite($file, $d[0]["value"]);
    }
    else {
      fwrite($file, $d);
    }
    $path = stream_get_meta_data($file)['uri'];
  }

  ob_start();
  passthru($exe . ' "--dolua:dofile([[posters.lua]]);attachfile([['.$path.']], [['. str_replace('"', $quote, $script).']], [['. $ext .']])"');
  $var = ob_get_contents();
  ob_end_clean();
  if ($file!=null)
    fclose($file);

  header('Access-Control-Allow-Origin: *');
  header('Content-Type: application/json');
  echo $var;
}

// post DICOM file, lua script determines what happens with it
// scripts in the form-data take preference over query string
function attachdicomfile($script) {
  include 'config.php';
  $file = null;

  if (array_key_exists('script', $_POST)) {
    $script = $_POST['script'];
  }

  if(count($_FILES) > 0){
    $first_file_index = $_FILES[array_keys($_FILES)[0]];
    $path = $first_file_index['tmp_name'];
  }
  else { 
    $d=file_get_contents("php://input");
    $t = getallheaders();
    if (!array_key_exists("Content-Type", $t)) {
      echo 'Does not contain content type';
      return;
    }
    $file = tmpfile();
    $typ=$t["Content-Type"];

    if (strpos($typ, 'multipart/related')!==false) {
      preg_match_all("/boundary=\"([^;= ]+)\"/", $typ, $r);
      $boundary= $r[1][0];

      $d = parse_multipart_content($d, $boundary);
      fwrite($file, $d[0]["value"]);
    }
    else {
      fwrite($file, $d);
    }
    $path = stream_get_meta_data($file)['uri'];
  }

  ob_start();
  passthru($exe . ' "--dolua:dofile([[posters.lua]]);attachdicomfile([['.$path.']], [['. str_replace('"', $quote, $script).']])"');
  $var = ob_get_contents();
  ob_end_clean();
  if ($file!=null)
    fclose($file);
  header('Access-Control-Allow-Origin: *');
  header('Content-Type: application/json');
  echo $var;
}

// post script to run immediately
function runscript() {
  include 'config.php';
  $script = $_POST['script'];

  $scr = tmpfile();
  fwrite($scr, $script);
  $scrpath = stream_get_meta_data($scr)['uri'];
  
  ob_start();
  passthru($exe . ' "--dolua:dofile([[posters.lua]]);runscript([['. $scrpath . ']])"');
  $var = ob_get_contents();
  ob_end_clean();
  fclose($scr);
  header('Access-Control-Allow-Origin: *');
  header('Content-Type: application/json');
  echo $var;
}

// post script to run in background; returns uid
function startscript() {
  include 'config.php';
  $script = $_POST['script'];

  $scr = tmpfile();
  fwrite($scr, $script);
  $scrpath = stream_get_meta_data($scr)['uri'];
  
  ob_start();
  passthru($exe . ' "--dolua:dofile([[posters.lua]]);startscript([['. $scrpath . ']])"');
  $var = ob_get_contents();
  ob_end_clean();
  fclose($scr);
  header('Access-Control-Allow-Origin: *');
  header('Content-Type: application/json');
  echo $var;
}

// read progress of script with given uid
function readprogress($uid) {
  include 'config.php';
  ob_start();
  passthru($exe . ' "--dolua:dofile([[posters.lua]]);readprogress([['. $uid . ']])"');
  $var = ob_get_contents();
  ob_end_clean();
  header('Access-Control-Allow-Origin: *');
  header('Content-Type: application/json');
  echo $var;
}

// write progress of script with given uid
function writeprogress($uid, $val) {
  include 'config.php';
  ob_start();
  passthru($exe . ' "--dolua:dofile([[posters.lua]]);writeprogress([['. $uid . ']], ' . $val . ')"');
  $var = ob_get_contents();
  ob_end_clean();
  header('Access-Control-Allow-Origin: *');
  header('Content-Type: application/json');
  echo $var;
}

File: webserver/htdocs/api/dicom/posters.lua
-- mvh 20230620 fix for linux (/ vs \)

function iowrite(a)
  if io then io.write(a)
  else write(a) end
end

-- add and process a file stored in path (may be zip)
-- script has importconverter format e.g. start with lua: if needed
function attachfile(path, script, ext)
  script = '[['..script..']]' -- allow ' and " in script
  ext = ext or ".dcm"
  remotecode = [[
    local script = ]]..script..[[;
    local dat=Command:GetVR(0x9999,0x0402,true);
    local s=Command["9999,0404"] or '0'
    if s=='1' then dat=string.sub(dat, 1, -2) end
    local filename=tempfile(']]..ext..[[');
    local f=io.open(filename,"wb");f:write(dat);f:close();
    servercommand('attachfile:'..filename..','..script)
    os.remove(filename);
  ]]
  iowrite(servercommand('lua:'..remotecode, '<'..path))
end

-- add and process a dicom file stored in path
-- script is in lua format working on global Data; return string taken as JSON response
function attachdicomfile(path, script)
  script = '[['..script..']]' -- allow ' and " in script
  ext = ".dcm"
  local remotecode = [[
    local script = ]]..script..[[;
    local dat=Command:GetVR(0x9999,0x0402,true);
    local s=Command["9999,0404"] or '0'
    if s=='1' then dat=string.sub(dat, 1, -2) end
    local filename=tempfile(']]..ext..[[');
    local f=io.open(filename,"wb");f:write(dat);f:close();
    readdicom(filename)
    os.remove(filename);
    return loadstring(script)()
  ]]
  iowrite(servercommand('lua:'..remotecode, '<'..path))
end

-- STOW a single dicom object stored in path; fix its PatientName and PatientID
function poststow(path)
  local code = [[
    local q=DicomObject:new()
    q.StudyInstanceUID=Data.StudyInstanceUID
    q.PatientID=""; q.PatientName=""
    local ae = servercommand('get_param:MyACRNema')
    r=dicomquery(ae, "STUDY", q)
    if r[0] then 
      Data.PatientID=r[0].PatientID
      Data.PatientName=r[0].PatientName 
    end;
    addimage(Data)
    local s='"'..Data.SOPInstanceUID..'"'
    if (#s % 2)~=0 then s=s..' ' end
    return s
  ]]
  attachdicomfile(path, code)
end

-- run a script in the server; can return information
function runscript(script)
  local remotecode = [[
    local dat=Command:GetVR(0x9999,0x0402,true);
    local s=Command["9999,0404"] or '0'
    if s=='1' then dat=string.sub(dat, 1, -2) end
    local filename=tempfile('.lua');
    local f=io.open(filename,"wb");f:write(dat);f:close();
    local a=dofile(filename);
    os.remove(filename);
    return a
  ]]
  iowrite(servercommand('lua:'..remotecode, '<'..script));
end

-- start a script as background task in the server; return uid of job
-- use progress(percentage) from script
function startscript(script)
  local remotecode1 = [[
    return tempfile('.lua');
  ]]
  local remotecode2 = [[
    local dat=Command:GetVR(0x9999,0x0402,true);
    local s=Command["9999,0404"] or '0'
    if s=='1' then dat=string.sub(dat, 1, -2) end
    local f=io.open(filename,"wb");f:write(dat);f:close();
    function progress(n) local f=io.open(filename, "w") f:write(n) f:close() end
    local a=dofile(filename);
    os.remove(filename);
    return a
  ]]
  local fn = servercommand('lua:'..remotecode1);
  local uid = '"' .. string.match(string.gsub(fn, '\\', '/'), '.+/(.-)%.lua') .. '"'
  iowrite(uid)
  servercommand('luastart:local filename=[['..fn..']];'..remotecode2, '<'..script);
end

-- return progress indicator of job (0 started, to 100 end)
function readprogress(uid)
  local remotecode = 
    "local uid=[["..uid.."]];" .. [[
    local fn = tempfile('.lua')
    local u1 = string.match(string.gsub(fn, '\\', '/'), '.+/(.-)%.lua')
    local filename = string.gsub(fn, u1, uid)
    local f = io.open(filename, 'r')
    if f==nil then
      return 100
    end
    local s = f:read('*a')
    f:close()
    return tonumber(s) or 0
  ]]
  local s = servercommand('lua:'..remotecode);
  iowrite(s)
end

-- write progress value of job (0 started, to 100 end)
function writeprogress(uid, val)
  local remotecode = 
    "local uid=[["..uid.."]];" ..
    "local val="..val..";" .. [[
    local fn = tempfile('.lua')
    local u1 = string.match(string.gsub(fn, '\\', '/'), '.+/(.-)%.lua')
    local filename = string.gsub(fn, u1, uid)
    local f = io.open(filename, 'w')
    f:write(val)
    f:close()
    return val
  ]]
  local s = servercommand('lua:'..remotecode);
  iowrite(s)
end
